libs/ml/tests/ml/ops/multiply.cpp

```diff
@@ -64,7 +64,125 @@ TYPED_TEST(MultiplyTest, forward_test)
+
+  // test correct values and shape
+  ASSERT_TRUE(prediction[0].AllClose(gt_1, fetch::math::function_tolerance<DataType>(),
+                                     fetch::math::function_tolerance<DataType>()));
+  ASSERT_TRUE(prediction[0].shape() == data_1.shape());
+  ASSERT_TRUE(prediction[1].AllClose(gt_2, fetch::math::function_tolerance<DataType>(),
+                                     fetch::math::function_tolerance<DataType>()));
+  ASSERT_TRUE(prediction[1].shape() == data_2.shape());
+}
+
+TYPED_TEST(MultiplyTest, backward_test_NMB_111)
+{
+  using DataType  = typename TypeParam::Type;
+  using ArrayType = TypeParam;
+
+  ArrayType data_1 = ArrayType::FromString(
+      "1, 2, 5, 6;"
+      "3, 4, 7, 8");
+  data_1.Reshape({2, 2, 2});
+
+  ArrayType data_2 = ArrayType::FromString("-1");
+  data_2.Reshape({1, 1, 1});
+
+  ArrayType error = ArrayType::FromString(
+      "0, 1, 4, 5;"
+      "2, 3, 6, 7");
+  error.Reshape({2, 2, 2});
+
+  ArrayType gt_1 = ArrayType::FromString(
+      "0, -1, -4, -5;"
+      "-2, -3, -6, -7");
+  gt_1.Reshape({2, 2, 2});
+
+  ArrayType gt_2 = ArrayType::FromString("168");
+  gt_2.Reshape({1, 1, 1});
+
+  fetch::ml::ops::Multiply<ArrayType> op;
+  std::vector<ArrayType>              prediction = op.Backward(
+      {std::make_shared<ArrayType>(data_1), std::make_shared<ArrayType>(data_2)}, error);
+
+  // test correct values and shape
+  ASSERT_TRUE(prediction[0].AllClose(gt_1, fetch::math::function_tolerance<DataType>(),
+                                     fetch::math::function_tolerance<DataType>()));
+  ASSERT_TRUE(prediction[0].shape() == data_1.shape());
+  ASSERT_TRUE(prediction[1].AllClose(gt_2, fetch::math::function_tolerance<DataType>(),
+                                     fetch::math::function_tolerance<DataType>()));
+  ASSERT_TRUE(prediction[1].shape() == data_2.shape());
+}
+
+TYPED_TEST(MultiplyTest, backward_test_NB_N1)
+{
+  using DataType  = typename TypeParam::Type;
+  using ArrayType = TypeParam;
+
+  ArrayType data_1 = ArrayType::FromString(
+      "1, 2, 5, 6;"
+      "3, 4, 7, 8");
+
+  ArrayType data_2 = ArrayType::FromString("1, -1");
+  data_2.Reshape({2, 1});
+
+  ArrayType error = ArrayType::FromString(
+      "0, 1, 4, 5;"
+      "2, 3, 6, 7");
+
+  ArrayType gt_1 = ArrayType::FromString(
+      "0, 1, 4, 5;"
+      "-2, -3, -6, -7");
+
+  ArrayType gt_2 = ArrayType::FromString(
+      "52;"
+      "116");
+  gt_2.Reshape({2, 1});
+
+  fetch::ml::ops::Multiply<ArrayType> op;
+  std::vector<ArrayType>              prediction = op.Backward(
+      {std::make_shared<ArrayType>(data_1), std::make_shared<ArrayType>(data_2)}, error);
+
+  // test correct values and shape
+  ASSERT_TRUE(prediction[0].AllClose(gt_1, fetch::math::function_tolerance<DataType>(),
+                                     fetch::math::function_tolerance<DataType>()));
+  ASSERT_TRUE(prediction[0].shape() == data_1.shape());
```

Comment:
You should test shape before AllClose, because AllClose would fail anyway, but it will give you a better idea of what is going on.